### PR TITLE
Refactor scheduler

### PR DESCRIPTION
**Pull request by jpbruinsslot**
Closes https://github.com/minvws/nl-kat-coordination/issues/1556

### CODE REVIEW

File: mula/scheduler/app.py [hunk 1]
```diff
@@ -223,28 +184,24 @@ def collect_metrics(self) -> None:
...
-        following processes:
-
-            * schedulers
-            * listeners
-            * monitors
-            * metrics collecting
-            * api server
+                following processes:
+
+        <<<<<<< HEAD
```
Comment (underdarknl) [inline mula/scheduler/app.py:189]: Merge conflict

**Comment by jpbruinsslot** [comment]
I anyone feels compelled to review then the PR is ready. I'm looking into how this could be separated into individual branches.

**Comment by praseodym** [comment]
When I try to run KAT from this branch, the scheduler gets into crash loop with the following log:
```
kat-scheduler-1  | WARNING:root:Deprecation: SCHEDULER_DB_DSN is deprecated, use SCHEDULER_DB_URI instead
kat-scheduler-1  | INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
kat-scheduler-1  | INFO  [alembic.runtime.migration] Will assume transactional DDL.
kat-scheduler-1  | WARNING:root:Deprecation: SCHEDULER_DB_DSN is deprecated, use SCHEDULER_DB_URI instead
kat-scheduler-1  | Traceback (most recent call last):
kat-scheduler-1  |   File "<frozen runpy>", line 198, in _run_module_as_main
kat-scheduler-1  |   File "<frozen runpy>", line 88, in _run_code
kat-scheduler-1  |   File "/app/scheduler/scheduler/__main__.py", line 6, in <module>
kat-scheduler-1  |     app = App(context.AppContext())
kat-scheduler-1  |               ^^^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/app/scheduler/scheduler/context/context.py", line 44, in __init__
kat-scheduler-1  |     katalogus_service = services.Katalogus(
kat-scheduler-1  |                         ^^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/katalogus.py", line 16, in __init__
kat-scheduler-1  |     super().__init__(host, source, timeout)
kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/services.py", line 83, in __init__
kat-scheduler-1  |     self._do_checks()
kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/services.py", line 158, in _do_checks
kat-scheduler-1  |     parsed_url = urllib.parse.urlparse(self.host)
kat-scheduler-1  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 394, in urlparse
kat-scheduler-1  |     url, scheme, _coerce_result = _coerce_args(url, scheme)
kat-scheduler-1  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 133, in _coerce_args
kat-scheduler-1  |     return _decode_args(args) + (_encode_result,)
kat-scheduler-1  |            ^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in _decode_args
kat-scheduler-1  |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
kat-scheduler-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in <genexpr>
kat-scheduler-1  |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
kat-scheduler-1  |                  ^^^^^^^^
kat-scheduler-1  | AttributeError: 'pydantic_core._pydantic_core.Url' object has no attribute 'decode'
```

**Comment by jpbruinsslot** [comment]
> When I try to run KAT from this branch, the scheduler gets into crash loop with the following log:
> 
> ```
> kat-scheduler-1  | WARNING:root:Deprecation: SCHEDULER_DB_DSN is deprecated, use SCHEDULER_DB_URI instead
> kat-scheduler-1  | INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
> kat-scheduler-1  | INFO  [alembic.runtime.migration] Will assume transactional DDL.
> kat-scheduler-1  | WARNING:root:Deprecation: SCHEDULER_DB_DSN is deprecated, use SCHEDULER_DB_URI instead
> kat-scheduler-1  | Traceback (most recent call last):
> kat-scheduler-1  |   File "<frozen runpy>", line 198, in _run_module_as_main
> kat-scheduler-1  |   File "<frozen runpy>", line 88, in _run_code
> kat-scheduler-1  |   File "/app/scheduler/scheduler/__main__.py", line 6, in <module>
> kat-scheduler-1  |     app = App(context.AppContext())
> kat-scheduler-1  |               ^^^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/app/scheduler/scheduler/context/context.py", line 44, in __init__
> kat-scheduler-1  |     katalogus_service = services.Katalogus(
> kat-scheduler-1  |                         ^^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/katalogus.py", line 16, in __init__
> kat-scheduler-1  |     super().__init__(host, source, timeout)
> kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/services.py", line 83, in __init__
> kat-scheduler-1  |     self._do_checks()
> kat-scheduler-1  |   File "/app/scheduler/scheduler/connectors/services/services.py", line 158, in _do_checks
> kat-scheduler-1  |     parsed_url = urllib.parse.urlparse(self.host)
> kat-scheduler-1  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 394, in urlparse
> kat-scheduler-1  |     url, scheme, _coerce_result = _coerce_args(url, scheme)
> kat-scheduler-1  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 133, in _coerce_args
> kat-scheduler-1  |     return _decode_args(args) + (_encode_result,)
> kat-scheduler-1  |            ^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in _decode_args
> kat-scheduler-1  |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
> kat-scheduler-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> kat-scheduler-1  |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in <genexpr>
> kat-scheduler-1  |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
> kat-scheduler-1  |                  ^^^^^^^^
> kat-scheduler-1  | AttributeError: 'pydantic_core._pydantic_core.Url' object has no attribute 'decode'
> kat-scheduler-1 exited with code 1
> ```

Thanks, I'll have a look.

**Comment by jpbruinsslot** [comment]
It probably has to do with the new pydantic version not being updated in your container.

**Comment by jpbruinsslot** [comment]
> > It probably has to do with the new pydantic version not being updated in your container.
> 
> The container is up to date:
> 
> ```
> > docker-compose run scheduler python
> [..]
> WARNING:root:Deprecation: SCHEDULER_DB_DSN is deprecated, use SCHEDULER_DB_URI instead
> INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
> INFO  [alembic.runtime.migration] Will assume transactional DDL.
> Python 3.11.5 (main, Aug 25 2023, 23:48:38) [GCC 12.2.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> from importlib.metadata import version
> >>> version('pydantic')
> '2.3.0'
> ```

Ok, I'll dig deeper

**Comment by praseodym** [comment]
FWIW, I'm seeing the same error in a clean GitHub Codespaces environment.

**Comment by jpbruinsslot** [comment]
Updated the code, had to do with pydantic and that the `Url` is being passed instead of being a string.

However I'm hitting a bytes authentication error at the moment. So I'll be trying to figure out what is going on there.

**Comment by jpbruinsslot** [comment]
@praseodym I solved the remaining issues, it should now be running correctly

**Comment by praseodym** [comment]
With OpenTelemetry (Jaeger) enabled I still see a similar error:

```
scheduler_1            | INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
scheduler_1            | INFO  [alembic.runtime.migration] Will assume transactional DDL.
scheduler_1            | [2023-08-28 14:47:05 +0000] [INFO] [Katalogus] Function is_host_available, executed successfully. Retry count: 0 [name=is_host_available, args=('katalogus', 8000), kwargs={}]
scheduler_1            | [2023-08-28 14:47:05 +0000] [INFO] [Katalogus] Function is_healthy, executed successfully. Retry count: 0 [name=is_healthy, args=(), kwargs={}]
scheduler_1            | [2023-08-28 14:47:06 +0000] [INFO] [Bytes] Function is_host_available, executed successfully. Retry count: 0 [name=is_host_available, args=('bytes', 8000), kwargs={}]
scheduler_1            | [2023-08-28 14:47:06 +0000] [INFO] [Bytes] Function is_healthy, executed successfully. Retry count: 0 [name=is_healthy, args=(), kwargs={}]
scheduler_1            | [2023-08-28 14:47:06 +0000] [INFO] [Octopoes] Function is_host_available, executed successfully. Retry count: 0 [name=is_host_available, args=('octopoes_api', 80), kwargs={}]
scheduler_1            | [2023-08-28 14:47:06 +0000] [INFO] [scheduler.server.server] Setting up instrumentation with span exporter endpoint [http://jaeger:4317/]
scheduler_1            | Traceback (most recent call last):
scheduler_1            |   File "<frozen runpy>", line 198, in _run_module_as_main
scheduler_1            |   File "<frozen runpy>", line 88, in _run_code
scheduler_1            |   File "/app/scheduler/scheduler/__main__.py", line 7, in <module>
scheduler_1            |     app.run()
scheduler_1            |   File "/app/scheduler/scheduler/app.py", line 218, in run
scheduler_1            |     self.server = server.Server(self.ctx, self.schedulers)
scheduler_1            |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/app/scheduler/scheduler/server/server.py", line 64, in __init__
scheduler_1            |     processor = BatchSpanProcessor(OTLPSpanExporter(endpoint=self.config.host_metrics))
scheduler_1            |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/usr/local/lib/python3.11/site-packages/opentelemetry/exporter/otlp/proto/grpc/trace_exporter/__init__.py", line 122, in __init__

scheduler_1            |     super().__init__(
scheduler_1            |   File "/usr/local/lib/python3.11/site-packages/opentelemetry/exporter/otlp/proto/grpc/exporter.py", line 185, in __init__
scheduler_1            |     parsed_url = urlparse(self._endpoint)
scheduler_1            |                  ^^^^^^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/usr/local/lib/python3.11/urllib/parse.py", line 394, in urlparse
scheduler_1            |     url, scheme, _coerce_result = _coerce_args(url, scheme)
scheduler_1            |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/usr/local/lib/python3.11/urllib/parse.py", line 133, in _coerce_args
scheduler_1            |     return _decode_args(args) + (_encode_result,)
scheduler_1            |            ^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in _decode_args
scheduler_1            |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
scheduler_1            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/usr/local/lib/python3.11/urllib/parse.py", line 117, in <genexpr>
scheduler_1            |     return tuple(x.decode(encoding, errors) if x else '' for x in args)
scheduler_1            |                  ^^^^^^^^
scheduler_1            | AttributeError: 'pydantic_core._pydantic_core.Url' object has no attribute 'decode'
scheduler_1            | [2023-08-28 14:47:06 +0000] [ERROR] [scheduler.utils.thread] Exception in thread: Thread-4 (push_tasks_for_random_objects)-scheduler-boefje-mispoes-random
scheduler_1            | Traceback (most recent call last):
scheduler_1            |   File "/app/scheduler/scheduler/utils/thread.py", line 65, in run_forever
scheduler_1            |     self._target()
scheduler_1            |   File "/usr/local/lib/python3.11/contextlib.py", line 81, in inner
scheduler_1            |     return func(*args, **kwds)
scheduler_1            |            ^^^^^^^^^^^^^^^^^^^
scheduler_1            |   File "/app/scheduler/scheduler/schedulers/boefje.py", line 311, in push_tasks_for_random_objects
scheduler_1            |     executor.submit(
scheduler_1            |   File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 169, in submit
scheduler_1            |     raise RuntimeError('cannot schedule new futures after '
scheduler_1            | RuntimeError: cannot schedule new futures after interpreter shutdown
scheduler_1            | [2023-08-28 14:47:06 +0000] [ERROR] [scheduler.app] Unhandled exception occurred: cannot schedule new futures after interpreter shutdown
```

**Review by praseodym — APPROVED** [review]
lgtm

**Comment by Darwinkel** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* No apparent regressions
* (Automatic) scheduling works
* Manually re-scheduling boefjes and normalizers works
* Downloading (running) boefje/normalizer data works

**Review by underdarknl — APPROVED** [review]

**Pull request merged**